### DISCUSSION TITLE
Proposal for Enhancing Code Quality and Contribution Guidelines

### DISCUSSION THREAD

**Discussion by fbalzarotti** [ImSwitch/ImSwitch#204] [Ideas]
# Goal
To ensure consistent code quality and streamline the development process, I propose the following changes related to 
- Code formatting
- Linting
- Contribution guidelines

These changes aim to establish clear definitions and enforcement mechanisms for code style and quality. 

# Action plan
## 1. Definitions and Enforcements

Steps:
- [ ] Call on active developers to explore alternatives/preferences on formatters and linters
- [ ] Find agreement via Github Discussions - avoid meeting(s)
  - We'll gather opinions here. If required, we'll make a poll.

Notes on the definitions to be made:

- **Formatter**. Define a standard code formatter to ensure consistent code style across the project.
- **Linter**. Define a standard linter to enforce code quality rules. This directly affects:
  - Documentation Generation: Ensure that docstrings adhere to a specific format for consistent documentation.
  - Type Hints: Enforce the use of type hints for better code readability and maintainability.
  - Import Sorting: Standardize import sorting to maintain clean and organized code.

## 2. Contribution Guidelines and Definition Files

Configure and document the usage of the formatter and linter in the project. Things to change:

- [ ] Contribution Guidelines. 
  Update the [contribution guidelines](https://github.com/ImSwitch/ImSwitch/blob/master/docs/contributing.rst) to include explicit instructions on:
  - Code formatting standards.
  - Linting requirements.
  - Procedures for ensuring compliance with these standards before submitting a pull request.
- [ ] Definition Files:
  - Add or update definition files (e.g., .toml files) in the root directory to enforce the specified linting and formatting rules.
  - These files will configure tools like ruff, black, flake8, isort, etc., ensuring that contributors have a clear and automated way to comply with the project’s coding standards.

## 3. Propagation
Update the upstream ImSwitch and its forks. 
Find a suitable strategy to do this painlessly (e.g. maybe it's necessary to make some merges before.)

# Timeline

I propose having an initial timeframe of ~2 weeks to retrieve opinions from active developers on step 1.

We'll adapt depending on the activity level.

**Comment by fbalzarotti** [comment]
Below we can gather information on the current uses in different groups like @beniroquai , @jacopoabramo , @GuillaumeMinet , @nornil , @kasasxav

**Reply by fbalzarotti** [reply]
In our group we use
- yapf as formatter https://github.com/google/yapf 
  - I just liked how the code looked better than black. But being so not mainstream, I don't think it is the right choice.
- pylint and isort because they come with VSCode

I quickly tried ruff https://docs.astral.sh/ruff/. As advertised, it is insanely fast and would be a good choice. 
It's a formatter and linter and it can be adapted by changing a .toml file.

**Reply by beniroquai** [reply]
Sinds very interesting @fbalzarotti. I'd say, let's stick to the community standard. I have absolutely no experience using Formatters, but it totally makes sense. @jacopoabramo which one are you using right now?

**Reply by nornil** [reply]
Thanks for bringing this up! I also have limited to no experience in formatters. However, I heard some use Pylama.. 
To my knowledge yapf is a pure formatter. How about linters?  Ruff seems promising based on my superficial github survey.

**Reply by fbalzarotti** [reply]
Thanks for the input. You will probably feel there's no way back once you use formatters and linters.
@Kubiczek36, @mgeismann might also have some thoughts.

**Reply by mgeismann** [reply]
+1 for ruff. Have been using it for a while for linting and I have nothing to complain about.

**Reply by beniroquai** [reply]
After review: ruff!

**Reply by Kubiczek36** [reply]
Checking Ruff and I like all in one solution configurable in the toml, which is transfered with the project. 🔥 Still need to try it though.

**Reply by xcasas** [reply]
In my current group we use black and pylint for style, bandit for security, pydocstyle for documentation, prospector for code analysis.

**Reply by GuillaumeMinet** [reply]
To my knowledge we don't use any formatter or linter in our group, and I have no experience using them on my side, so no opinion from me.
We'll be happy to implement whatever is decided here, of course.
